Bolts/Common/BFTask.m

```diff
@@ -418,16 +455,16 @@ - (void)waitUntilFinished {
     if ([NSThread isMainThread]) {
         [self warnOperationOnMainThread];
     }
-
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return;
+    BOOL completed = self.isCompleted;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        if (!completed) {
+            [self.condition lock];
```

Comment:
This is technically undefined behavior here. There's no guarantee on the thread that synchronizationQueue runs on, and NSConditionLock requires you to unlock from an identical thread. We may need to rethink this. (What's the bolts-swift way of doing this?)

Reply:
seems very similar: https://github.com/BoltsFramework/Bolts-Swift/blob/master/Sources/BoltsSwift/Task.swift#L211

main difference is that the NSConditionLock is created set on the queue, let me know if you prefer that way.